### ISSUE TITLE
Shown song length is 0:00 when a song is transcoded

### ISSUE THREAD

**Issue by Phyks** [ampache/ampache#1303]
# Environment
- Ampache version: develop
- Web server + version:
- Server operating system:
- Client operating system:

- Client type:
  - Web
# Settings

# Description

- [ ] When a played song is transcoded, many features do not work including the shown song length, which is always 0:00.

This is also related to https://github.com/ampache/ampache/issues/1048 and other transcoding issues.
# Logs

**Comment by Niols** [comment]
I don't know if it is an issue in Ampache or in the way Ampache uses FFMPEG.
Anyway, we need to find a way to send the metadata with the transcoded file

**Comment by Afterster** [comment]
This one is outside Ampache scope unfortunately (it was already discussed several time).
I agree with @Niols about its suggestion.

**Comment by Phyks** [comment]
I agree this one has already been discussed, I saw many references to it, but never explicitly (but I may have missed an issue).

This one also has side effects such as clicking "play" at the end of the file may play beyond the file end and related things.

**Comment by Phyks** [comment]
- [ ] Also, when forced transcoding is on, the repeat button does not work.

**Comment by Psy-Virus** [comment]
I never got any reply to [my two cents](https://github.com/ampache/ampache/issues/1048#issuecomment-167266289) on #1048. I know, it's some kinda cheating, but it would help until there's some kind of an official way(?)
"A compromise is an agreement, with which no party is happy"^^

**Comment by Phyks** [comment]
Sorry, I missed it and actually reported the issue, but not had time to have a look at it deeper and ways to tackle it.

This looks like a good compromise. However, I am not convinced that checking the `ffmpeg` output in a pipe would be much more difficult to do. I should have a deeper look at it though.

**Comment by Psy-Virus** [comment]
Could this help? It's also from libavtools

**Comment by Psy-Virus** [comment]
Could this help? It's also from libavtools:

```
$ avprobe testfile\ -\ testfile.mp3
avprobe version 9.18-6:9.18-0ubuntu0.14.04.1, Copyright (c) 2007-2014 the Libav developers
  built on Mar 16 2015 13:19:10 with gcc 4.8 (Ubuntu 4.8.2-19ubuntu1)
[mp3 @ 0x224c320] max_analyze_duration reached
[mp3 @ 0x224c320] Estimating duration from bitrate, this may be inaccurate                                                                    
Input #0, mp3, from 'testfile - testfile.mp3':
  Metadata:
    album           : testalbum
    title           : testfile
    album_artist    : testinterpret
    performer       : testdirigenten
    disc            : dies ist ein testfile
    TKEY            : einer der 13 des kleinen gespensts
    TIT1            : gruppe für testfiles
    publisher       : testherausgeber
    encoded_by      : testcodec
    track           : 42
    date            : 42
    genre           : testgenre
    composer        : testkomponisten
    artist          : testinterpret
    replaygain_reference_loudness: 89.0 dB
    replaygain_track_gain: -2.640000 dB
    replaygain_track_peak: 0.606325
    MP3GAIN_MINMAX  : 059,207
    MP3GAIN_UNDO    : +003,+003,N
  Duration: 00:02:33.04, start: 0.000000, bitrate: 320 kb/s
    Stream #0.0: Audio: mp3, 44100 Hz, 2 channels, s16p, 320 kb/s
# avprobe output
```

**Comment by Psy-Virus** [comment]
Sorry, ctrl+enter -.-

**Comment by Psy-Virus** [comment]
You could simply grep the length/duration. Or in a PHP way

**Comment by Phyks** [comment]
Sure at the length/duration. We should already have it in db anyway.

Actually, when you posted your message, I was reading [this article](https://linuxfr.org/news/fwomaj-0-3-videos-a-la-coupe-au-rayon-frais) (sorry, in French):

> Du coup, comme je ne lis pas la sortie de FFmpeg (je vérifie juste les codes de retour) je suis en asynchrone, yay, mais je perds la possibilité d'informer l'utilisateur de l'avancement du travail en cours, bouh ; juste un bidule qui pulse pour indiquer qu'il se passe quelque chose. Notez que FFmpeg lui-même ne le fait pas, mais il indique quelle image il est en train de traiter, or moi, j'ai plein de moyens de savoir combien d'images fait mon fichier, à commencer par Mediainfo. Eh. Dommage.

The author seems to state that FFmpeg offers a way to get the current frame it is handling. As we have the total number of frames as well, we can show progressbars and stuff like this with a bit of wrapping I guess. Still, this is purely hypothetical and I should have a deeper look at how FFmpeg is plugged in Ampache.

**Comment by Psy-Virus** [comment]
> Sure at the length/duration. We should already have it in db anyway.

Hm ehm yeah.... Now I'm feeling stupid, Thanks! XDD

But it sounds interesting. If it's possible, we won't need any dirty fix.

**Comment by Psy-Virus** [comment]
Just something that came across my head:
If it isn't really possible get the Info from the stream; how would it be to calculate the stream/file size divided by the bit rate? I mean, the source files are transcoded so there must be something that has a file size. And we know the bit rate of the stream, or it can be read out.
So if an mp3 has a file size of 3.5 MB and a bit rate of 128 kbit/s; we have a file length of 229.376 seconds (3.5 x 1024² / 128000 = seconds). This is a result with milliseconds but absolutely theoretical because the file size will differ a small amount if there is any TAG information (picture, artist name, etc.). So this should also be standardized or simply deleted from the stream while transcoding. so for the total length, the database information could be used and for a "loading bar" the calculated results could be used.

**Comment by Phyks** [comment]
Yes, I am pretty much sure we can come up with some hacky thing to
handle this. Will try to find some time to give it a try and make a PR.

I'd like to see how it goes exactly and if we can make some exact
computation.
-- 
Phyks

**Comment by Niols** [comment]
Hum, what about files with variable bitrates?

**Comment by Psy-Virus** [comment]
Do you mean the source or transcoded files?
I don't think, that it will work on the transcoded files, because there will be calculations that are maybe done while the fill is creating. But for the source file, avprobe for example delivers the average bit rate. 

Look at the `Stream #0.0:` thing

```text
$ avprobe -analyzeduration 1000000000 "testfilevariablebr - testfilevariablebr.mp3" 
avprobe version 9.18-6:9.18-0ubuntu0.14.04.1, Copyright (c) 2007-2014 the Libav developers
  built on Mar 16 2015 13:19:10 with gcc 4.8 (Ubuntu 4.8.2-19ubuntu1)
Input #0, mp3, from 'testfilevariablebr - testfilevariablebr.mp3':
  Metadata:
    album           : testalbum
    artist          : testfilevariablebr
    album_artist    : testfilevariablebr
    title           : testfilevariablebr
  Duration: 00:03:09.28, start: 0.000000, bitrate: 212 kb/s
    Stream #0.0: Audio: mp3, 44100 Hz, 2 channels, s16p, 211 kb/s
# avprobe output
```

**Comment by Phyks** [comment]
Weird thing we just noticed:
Try to play a song (typically FLAC) with transcoding set to "always" (MP3, 128kbps as output) in Firefox or Chromium, song length is "0:00" and waveform does not work, as described above.
Now, try to do the same with Edge (latest version), everything is working fine.

Dev tools does not say anything in Firefox and Chromium. I do not really know what is happening. My best bet is that somehow Edge waits for the full stream to be available and download the transcoded file, rather than streaming the file directly as soon as it is available as Firefox/Chromium do. But this is only a supposition.

cc @Gano16 who found this.

**Comment by GeoTob** [comment]
@Phyks re your last comment and edit above: do you have an idea what's causing the the player refresh on Edge or what's preventing it on Chromium/FF? Would this be coming from ampache or jplayer?

**Comment by lachlan-00** [comment]
database time is sent to the player without worrying about the file when transcoded so should be fine now
